File: 1501~2000/1578. Minimum Time to Make Rope Colorful/main.py
"""
Source: https://leetcode.com/problems/minimum-time-to-make-rope-colorful/
Date: 2023/4/18
Skill:
Ref:
Runtime: 1184 ms, faster than 50.27%
Memory Usage: 22.7 MB, less than 99.22%
Time complexity:
Space complexity:
Constraints:

"""

import math
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right


class Solution:
    def minCost(self, colors: str, neededTime: List[int]) -> int:
        sz, res = len(colors), 0
        for start in range(sz - 1):
            if colors[start] == colors[start + 1]:
                res += min(neededTime[start], neededTime[start + 1])
                neededTime[start + 1] = max(neededTime[start], neededTime[start + 1])

        return res


if __name__ == "__main__":
    s = Solution()

File: 1501~2000/1578. Minimum Time to Make Rope Colorful/main.cpp
/**
 * Source: is.gd/RHgs9D
 * Date: 2023/12/27
 * Skill:
 * Ref:
 * Runtime: 115 ms, faster than 70.49% of C++ online submissions
 * Memory Usage: 95.70 MB, less than 83.31% of C++ online submissions
 * Time complexity:
 * Space complexity:
 * Constraints:
 *
 */


#include <iostream>
#include <vector>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>
#include <functional>
#include <list>


using namespace std;
#define SZ(X) ((int)(X).size())
const int MX = 1e5+1;


class Solution {
public:
    int minCost(string colors, vector<int>& neededTime) {
        int res = 0, len = SZ(neededTime);
        for (int i = 0; i < len; i++) {
            int j = i, total = 0, mx = 0;
            while (j < len && colors[j] == colors[i]) {
                total += neededTime[j];
                mx = max(mx, neededTime[j++]);
            }
            res += total - mx;
            i = j - 1;
        }
        return res;
    }
};